cdk/lib/recipes-reindex.ts

```diff
@@ -0,0 +1,182 @@
+		const writeBatchToReindexQueue = new GuLambdaFunction(
+			scope,
+			'WriteBatchToReindexQueueLambda',
+			{
+				app: `${appBase}-write-batch-to-index-queue`,
+				description: 'Write a batch of recipe ids to the reindex queue',
+				fileName: lambdaFileName,
+				handler: 'main.writeBatchToReindexQueueHandler',
+				functionName: `${appBase}-write-batch-to-reindex-queue-${scope.stage}`,
+				runtime: Runtime.NODEJS_20_X,
+				initialPolicy: [
+					new PolicyStatement({
+						effect: Effect.ALLOW,
+						actions: ['s3:GetObject'],
+						resources: [snapshotBucket.bucketArn + '/*'],
+					}),
+				],
+				environment: {
+					RECIPE_INDEX_SNAPSHOT_BUCKET: snapshotBucket.bucketName,
+					REINDEX_BATCH_SIZE: reindexBatchSize.toString(),
+				},
+				architecture: Architecture.ARM_64,
+				timeout: Duration.seconds(30),
+				memorySize: 128,
+			},
+		);
+
+		// Define the Step Functions tasks
+		const checkForOtherRunningReindexesTask = new CustomState(
+			scope,
+			'checkForOtherRunningTasks',
+			{
+				stateJson: {
+					Type: 'Task',
+					Parameters: {
+						StatusFilter: 'RUNNING',
+						// The ARN of the running state machine
+						'StateMachineArn.$': '$$.StateMachine.Id',
+					},
+					Resource: 'arn:aws:states:::aws-sdk:sfn:listExecutions',
+				},
+			},
+		);
+
+		const snapshotOrderedIndexTask = new LambdaInvoke(
+			this,
+			'SnapshotOrderedIndexTask',
+			{
+				lambdaFunction: snapshotRecipeIndex,
+				payload: TaskInput.fromObject({
+					'executionId.$': '$$.Execution.Name',
+				}),
+			},
+		);
+
+		const writeBatchToReindexQueueTask = new LambdaInvoke(
+			this,
+			'WriteBatchToReindexQueueTask',
+			{
+				lambdaFunction: writeBatchToReindexQueue,
+				inputPath: '$.Payload',
+			},
+		);
+
+		const waitForThroughputAndWriteNextBatch = new Wait(
+			scope,
+			'WaitForThroughPut',
+			{
+				time: WaitTime.duration(Duration.seconds(1)),
+			},
+		).next(writeBatchToReindexQueueTask);
+
+		const isReindexComplete = new Choice(this, 'IsReindexComplete')
+			.when(
+				Condition.numberGreaterThanJsonPath(
+					'$.Payload.currentIndex',
+					'$.Payload.lastIndex',
+				),
+				new Succeed(this, 'Reindex complete'),
+			)
+			.otherwise(waitForThroughputAndWriteNextBatch);
+
+		snapshotOrderedIndexTask
+			.next(writeBatchToReindexQueueTask)
+			.next(isReindexComplete);
+
+		const isOnlyRunningReindex = new Choice(this, 'IsOnlyRunningReindex')
+			.when(Condition.isNotPresent('$.Executions[1]'), snapshotOrderedIndexTask)
+			.otherwise(
+				new Fail(this, 'Fail', {
+					error: 'ReindexingError',
+					cause: 'Other reindexes running',
+				}),
+			);
+
+		// Define the state machine
+		const definition =
+			checkForOtherRunningReindexesTask.next(isOnlyRunningReindex);
+
+		const stateMachine = new StateMachine(this, 'ReindexingStateMachine', {
+			definition,
+			timeout: Duration.minutes(15),
+		});
+
+		stateMachine.addToRolePolicy(
+			new PolicyStatement({
+				effect: Effect.ALLOW,
+				resources: ['*'],
```

Comment:
I don't suppose we can narrow this down at all?

Reply:
Such a good question. I tried, good lord I tried, to scope this to the SM in question, and could not avoid CDK complaining about a circular dependency.

Having said that, I've just re-searched our estate and it looks like it might be possible to do this by reconstructing the arn manually — bear with ...

Reply:
Ironically your other comment https://github.com/guardian/recipes-backend/pull/79#issuecomment-2515093595 is the solution to this — by manually defining a name and manually constructing the result ARN we can avoid a forward reference 👍  dd5ece910680dc7f337fdc1fac58f16e9b00a9ee